dipy/denoise/denspeed.pyx

```diff
@@ -153,49 +157,70 @@ cdef double process_block(double [:, :, ::1] arr,
 
     cdef:
         cnp.npy_intp m, n, o, M, N, O, a, b, c, cnt, step
-        double patch_vol_size
+        double patch_vol_size, block_vol_size
         double summ, d, w, sumw, sum_out, x, sigm
         double * W
         double * cache
         double * sigma_block
         double denom
         cnp.npy_intp BS = B * 2 + 1
+        cnp.npy_intp PS = P * 2 + 1
+        cnp.npy_intp TS = PS + BS - 1
 
     cnt = 0
     sumw = 0
     patch_vol_size = (P + P + 1) * (P + P + 1) * (P + P + 1)
+    block_vol_size = (B + B + 1) * (B + B + 1) * (B + B + 1)
 
-    W = <double *> malloc(BS * BS * BS * sizeof(double))
-    cache = <double *> malloc(BS * BS * BS * sizeof(double))
-    sigma_block = <double *> malloc(BS * BS * BS * sizeof(double))
+    W = <double *> malloc(PS * PS * PS * sizeof(double))
+    cache = <double *> malloc(TS * TS * TS * sizeof(double))
+    sigma_block = <double *> malloc(TS * TS * TS * sizeof(double))
 
     # (i, j, k) coordinates are the center of the static patch
     # copy block in cache
-    copy_block_3d(cache, BS, BS, BS, arr, i - B, j - B, k - B)
-    copy_block_3d(sigma_block, BS, BS, BS, sigma, i - B, j - B, k - B)
+    copy_block_3d(
+        cache,
+        PS + BS - 1,
+        PS + BS - 1,
+        PS + BS - 1,
+        arr,
+        i - P - B,
+        j - P - B,
+        k - P - B)
+    copy_block_3d(
+        sigma_block,
+        PS + BS - 1,
+        PS + BS - 1,
+        PS + BS - 1,
+        arr,
+        i - P - B,
+        j - P - B,
+        k - P - B)
 
     # calculate weights between the central patch and the moving patch in block
     # (m, n, o) coordinates are the center of the moving patch
     # (a, b, c) run inside both patches
-    for m in range(P, BS - P):
-        for n in range(P, BS - P):
-            for o in range(P, BS - P):
```

Comment:
@Garyfallidis @omarocegueda 
The essential algorithm following [this](http://link.springer.com/chapter/10.1007%2F978-3-540-85990-1_15) is that we have a patch(V) around a voxel(x) of radius (P), and then we do the following to get the denoised output
<img width="458" alt="screen shot 2016-06-22 at 11 19 56 pm" src="https://cloud.githubusercontent.com/assets/7941975/16277324/121696b4-38d0-11e6-8a75-81e56092d7eb.png">

and the weights are computed by computing distance between a patch (size S) with a different radius (B), centered at x and x_i, for all i in V. Mathematically 

<img width="417" alt="screen shot 2016-06-22 at 11 28 25 pm" src="https://cloud.githubusercontent.com/assets/7941975/16277558/126fbf18-38d1-11e6-8cd2-17484cba8726.png">

So in the current code there is some correction in normalization of the weights (will correct that, but it does not affect the oversmoothing problem). 

The changed code exactly follows the algorithm. The problem with the previous code lies primarily in these lines as it did something entirely different, and if you keep P=B  or P>B you can see these loops will cause a problem. Hence I restructured it totally and followed the paper's algorithm step by step. I rechecked it, I do not find any problem with the implementation apart from the normalization of sigma (lines 222, 223). Please do have look at this.